vpr/src/place/analytical_placement/AnalyticalSolver.cpp

```diff
@@ -162,21 +211,255 @@ void QPHybridSolver::solve(unsigned iteration, PartialPlacement &p_placement) {
+            if (!p_placement.is_moveable_node(first_node_id)) {
+                if (!p_placement.is_moveable_node(second_node_id)) {
+                    return;
+                }
+                std::swap(first_node_id, second_node_id);
+            }
+            double epsilon = 1e-6;
+            double dist = 0; 
+            if(is_x)
+                dist = 1.0/std::max(std::abs(p_placement.node_loc_x[first_node_id] - p_placement.node_loc_x[second_node_id]), epsilon);
+            else
+                dist = 1.0/std::max(std::abs(p_placement.node_loc_y[first_node_id] - p_placement.node_loc_y[second_node_id]), epsilon);
+            double w = (2.0/static_cast<double>(num_nodes - 1)) * dist;
+            if (p_placement.is_moveable_node(second_node_id)) {
+                if (is_x) {
+                    tripletList_x.emplace_back(first_node_id, first_node_id, w);
+                    tripletList_x.emplace_back(second_node_id, second_node_id, w);
+                    tripletList_x.emplace_back(first_node_id, second_node_id, -w);
+                    tripletList_x.emplace_back(second_node_id, first_node_id, -w);
+                }else{
+                    tripletList_y.emplace_back(first_node_id, first_node_id, w);
+                    tripletList_y.emplace_back(second_node_id, second_node_id, w);
+                    tripletList_y.emplace_back(first_node_id, second_node_id, -w);
+                    tripletList_y.emplace_back(second_node_id, first_node_id, -w);
+                }
+            } else {
+                if (is_x){
+                    tripletList_x.emplace_back(first_node_id, first_node_id, w);
+                    b_x(first_node_id) += w * p_placement.node_loc_x[second_node_id];
+                }else{
+                    tripletList_y.emplace_back(first_node_id, first_node_id, w);
+                    b_y(first_node_id) += w * p_placement.node_loc_y[second_node_id];
+                }
+            }
+        };
+        for (size_t node_id = 0; node_id < node_ids_no_duplicate.size(); node_id++) {
+            if (node_id != maxXId && node_id != minXId) {
+                add_node(node_id, maxXId, node_ids_no_duplicate.size(), true);
+                add_node(node_id, minXId, node_ids_no_duplicate.size(), true);
+            } 
+            if (node_id != maxYId && node_id != minYId) {
+                add_node(node_id, maxYId, node_ids_no_duplicate.size(), false);
+                add_node(node_id, minYId, node_ids_no_duplicate.size(), false);
+            } 
+        }
+        add_node(maxXId, minXId, node_ids_no_duplicate.size(), true);
+        add_node(maxYId, minYId, node_ids_no_duplicate.size(), false);
+    }
+    A_sparse_x.setFromTriplets(tripletList_x.begin(), tripletList_x.end());
+    A_sparse_y.setFromTriplets(tripletList_y.begin(), tripletList_y.end());
+}
+
+
+void B2BSolver::populate_matrix_anchor(PartialPlacement& p_placement, unsigned iteration) {
+    double epsilon = 1e-6;
```

Comment:
This epsilon is used in a couple of places. You should make it a static const member which will have 0 performance implications. I would declare it `inline static const epsilon = 1e-16;` as a data member in the class itself. In the future we would want this to be user-controlled.